### PR TITLE
RES-239 Add EXPoSE algorithm detector to NAB

### PR DESCRIPTION
**Pull request by zuhaagha**
EXPoSE algorithm with decay implementation based on https://arxiv.org/abs/1601.06602.
After tuning, the final scores are close to sliding window and change point. Tuning was done on domino and parameters varied include `gamma`, `decay`, `randomSeed` and `n_components`.

### CODE REVIEW

**Comment by zuhaagha** [comment]
@BoltzmannBrain @subutai please review

File: nab/detectors/expose/expose_detector.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    self.decay = 0.01
+    self.timestep = 1.0
+
+    # Parameters for RBF kernel initialization
+    self.n_components = 20000 #dimensionality of feature transform
+    self.gamma = 0.5
+    self.randomSeed = 290
+
+
+  def initialize(self):
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: Why not do this is the constructor? Then `n_components`, `gamma`, and `randomSeed` don't need to be maintained.

File: nab/detectors/expose/expose_detector.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  data point to previous points without assuming an underlying data
+  distribution.
+
+  There are three EXPoSE variants : incremental, windowing and decay. This
+  implementation is based on EXPoSE with decay. All three variants have been
+  tried on NAB but decay gives the best results.
+
+  Parameters for for this detector have been tuned to give the best
+  performance on NAB. Anomaly scores are in the range of -0.02 to 1.02
+  because of approximation.
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: Approximation? What/where?

File: nab/detectors/expose/expose_detector.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class ExposeDetector(AnomalyDetector):
+
+  """ This detector is an implemetation of The EXPoSE (EXPected Similarity
+  Estimation) algorithm as described in Markus Schneider, Wolfgang Ertel,
+  Fabio Ramos, "Expected Similarity Estimation for Lage-Scale Batch and
+  Streaming Anomaly Detection", arXiv 1601.06602 (2016).
+
+  EXPose is an anomaly detection classifier that calculates the likelihood of a
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: This is a long, difficult to understand sentence.

Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: EXPoSE

File: nab/detectors/expose/expose_detector.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import math
+import numpy
+from sklearn.kernel_approximation import RBFSampler
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: blank line between built-in and external module imports

File: README.md [hunk 5]
```diff
@@ -38,6 +38,7 @@ The NAB scores are normalized such that the maximum possible is 100.0 (i.e. the
 | [Twitter ADVec v1.0.0](https://github.com/twitter/AnomalyDetection)| 47.1             | 33.6          | 53.5          |
 | [Etsy Skyline](https://github.com/etsy/skyline) | 35.7             | 27.1          | 44.5          |
 | Bayesian Changepoint**          | 17.7              | 3.2           | 32.2           |
+| [EXPoSE] (https://github.com/numenta/NAB/blob/master/nab/detectors/expose/expose_detector.py)    | 16.44     | 3.19  | 26.92     |
```
Comment (BoltzmannBrain) [inline README.md]: please be consistent with decimal places

File: nab/detectors/expose/expose_detector.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    self.randomSeed = 290
+
+
+  def initialize(self):
+    self.kernel = RBFSampler(gamma=self.gamma,
+                             n_components=self.n_components,
+                             random_state=self.randomSeed)
+
+
+  def handleRecord(self, inputData):
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: please add docstring

File: nab/detectors/expose/expose_detector.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    # Update previous expose model
+    self.previousExposeModel = exposeModel
+
+    # Compute anomaly score by calculating similarity of the new data point
+    # with expose model. The similarity measure, calculated via inner
+    # product, is the likelihood of the data point being normal.
+    anomalyScore = numpy.asscalar(1 - numpy.inner(inputFeature, exposeModel))
+    self.timestep += 1
+
+    return [anomalyScore]
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: blank line at the end of Python files

File: nab/detectors/expose/expose_detector.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  def initialize(self):
+    self.kernel = RBFSampler(gamma=self.gamma,
+                             n_components=self.n_components,
+                             random_state=self.randomSeed)
+
+
+  def handleRecord(self, inputData):
+    # Transform the input by approximating feature map of a Radial Basis
+    # Function kernel using Random Kitchen Sinks approximation
+    inputFeature = self.kernel.fit_transform(numpy.array([[inputData[
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: cleaner to break the line at the function args: `inputFeature = self.kernel.fit_transform(...`

File: nab/detectors/expose/expose_detector.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                                                             "value"]]]))
+
+    # Compute expose model as a weighted sum of new data point's feature
+    # map and previous data points' kernel embedding. Influence of older data
+    # points declines with the decay factor.
+    if self.timestep == 1.0:
+      exposeModel = inputFeature
+    else:
+      exposeModel = (self.decay * inputFeature) + (1 - self.decay) * \
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: break lines with implicit continuation of parenthesis, not backslash

File: nab/detectors/expose/expose_detector.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  because of approximation.
+  """
+
+  def __init__(self, *args, **kwargs):
+    super(ExposeDetector, self).__init__(*args, **kwargs)
+
+    self.kernel = None
+    self.previousExposeModel = []
+    self.decay = 0.01
+    self.timestep = 1.0
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: Why float?

**Comment by BoltzmannBrain** [comment]
@zuhaagha done with this round of review. You'll also need to edit run.py to enable this detector. This detector uses sklearn, a module we currently don't require in NAB and probably shouldn't. In the run.py edit you'll make, there should a comment stating that in order to run this detector you must have the sklearn package.

**Comment by zuhaagha** [comment]
Thanks @BoltzmannBrain! I am looking into the changes. And just realized I forgot to add run.py when committing changes. Thanks for catching that!

File: nab/detectors/expose/expose_detector.py (hunk 10 above)
Comment (zuhaagha) [inline nab/detectors/expose/expose_detector.py]: I modified it to float for division when implementing windowing approach. But float is unneeded here, so I'll change it.

**Comment by zuhaagha** [comment]
@BoltzmannBrain review changes made

File: nab/detectors/expose/expose_detector.py (hunk 2 above)
Comment (zuhaagha) [inline nab/detectors/expose/expose_detector.py]: RBF kitchen sink approximation. But I just got rid of that phrase :)

File: nab/detectors/expose/expose_detector.py [hunk 11]
```diff
@@ -1,33 +1,28 @@
-import math
 import numpy
-from sklearn.kernel_approximation import RBFSampler
-from sklearn.preprocessing import normalize
 
+from sklearn.kernel_approximation import RBFSampler
 from nab.detectors.base import AnomalyDetector
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]:
```
import numpy

from sklearn.kernel_approximation import RBFSampler

from nab.detectors.base import AnomalyDetector
```

File: README.md [hunk 12]
```diff
@@ -38,6 +38,7 @@ The NAB scores are normalized such that the maximum possible is 100.0 (i.e. the
 | [Twitter ADVec v1.0.0](https://github.com/twitter/AnomalyDetection)| 47.1             | 33.6          | 53.5          |
 | [Etsy Skyline](https://github.com/etsy/skyline) | 35.7             | 27.1          | 44.5          |
 | Bayesian Changepoint**          | 17.7              | 3.2           | 32.2           |
+|  EXPoSE   | 16.4     | 3.2  | 26.9     |
```
Comment (BoltzmannBrain) [inline README.md]: link to this http://arxiv.org/abs/1511.05371

File: nab/detectors/expose/expose_detector.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
+import numpy
+
+from sklearn.kernel_approximation import RBFSampler
+from nab.detectors.base import AnomalyDetector
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]:
```
import numpy

from sklearn.kernel_approximation import RBFSampler

from nab.detectors.base import AnomalyDetector
```

File: nab/detectors/expose/expose_detector.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  the inner product of its feature map with kernel embedding of previous data
+  points. This measures the similarity of a data point to previous points
+  without assuming an underlying data distribution.
+
+  There are three EXPoSE variants: incremental, windowing and decay. This
+  implementation is based on EXPoSE with decay. All three variants have been
+  tried on NAB but decay gives the best results.
+
+  Parameters for this detector have been tuned to give the best performance.
+  Anomaly score outputs are in the range of -0.02 to 1.02.
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: A specific comment like this would go in `handleRecord()`

**Comment by BoltzmannBrain** [comment]
@zuhaagha this doesn't run for me, I think b/c I have numpy 1.9.2 (as req'd in nupic). What are your versions:
```
>>> numpy.__version__
'1.9.2'
>>> pandas.__version__
u'0.18.1'
>>> scipy.__version__
'0.13.0b1'
```

**Comment by zuhaagha** [comment]
I have 1.11.0. Must have upgraded my installation for sklearn. Let me try it with 1.9.2.

File: README.md (hunk 12 above)
Comment (zuhaagha) [inline README.md]: Should it be linked to this one or the original expose paper https://arxiv.org/pdf/1601.06602v3.pdf? The implementation is for the original paper w/o stochastic optimization.

Comment (BoltzmannBrain) [inline README.md]: Oh good catch. Link it here please: https://arxiv.org/abs/1601.06602v3

**Comment by zuhaagha** [comment]
I think it's the scipy version and not numpy version which might be causing the problem. My scipy version is 0.17.1

**Comment by zuhaagha** [comment]
@BoltzmannBrain are you getting the error below,

`ValueError: numpy.dtype has the wrong size, try recompiling`

**Comment by BoltzmannBrain** [comment]
yes

**Comment by zuhaagha** [comment]
@BoltzmannBrain To make it run with numpy 1.9.2, 
pip uninstall sklearn (i.e the latest version 0.17.1) 
pip install sklearn==0.16.1 --user

The latest version of sklearn runs with numpy 1.11.0 or higher. Older version works with numpy 1.9.2. Let me know if it works now and i'll update run.py with the specific version of sklearn in comments.

**Comment by BoltzmannBrain** [comment]
Thanks, but it would be `scikit-learn` instead of `sklearn` 😉 . I'm running it now...

**Comment by zuhaagha** [comment]
Oh yes! Typo...

File: nab/detectors/expose/expose_detector.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  have been tuned to give the best performance.
+  """
+
+  def __init__(self, *args, **kwargs):
+    super(ExposeDetector, self).__init__(*args, **kwargs)
+
+    self.kernel = None
+    self.previousExposeModel = []
+    self.decay = 0.01
+    self.timestep = 1
```
Comment (BoltzmannBrain) [inline nab/detectors/expose/expose_detector.py]: Convention is to index from 0, so please start `self.timestep` from 0 and modify the conditional at line 55

**Comment by BoltzmannBrain** [comment]
@zuhaagha left one last comment, then 👍

**Comment by zuhaagha** [comment]
@BoltzmannBrain Done! Will merge after travis passes

**Pull request merged**